File: 77-combinations/solution.py
class Solution:
  @staticmethod
  def combine_helper(start: int, n: int, k: int, builder: list[int], result: list[list[int]]) -> None:
    if len(builder) == k:
      result.append(builder.copy())
      return

    needed = k - len(builder)
    if start + needed - 1 > n:
      return

    for i in range(start, n + 1):
      builder.append(i)
      Solution.combine_helper(i + 1, n, k, builder, result)
      builder.pop()

  def combine(self, n: int, k: int) -> list[list[int]]:
    result: list[list[int]] = []
    Solution.combine_helper(1, n, k, [], result)
    return result

File: 77-combinations/solution.cpp
#include <vector>

class Solution
{
private:
  static void combine_helper(const int start, const int n, const int k, std::vector<int> &builder, std::vector<std::vector<int>> &result)
  {
    if (builder.size() == k)
    {
      result.emplace_back(builder);
      return;
    }

    const int needed = k - builder.size();
    if (start + needed - 1 > n)
    {
      return;
    }

    for (int i = start; i < n + 1; ++i)
    {
      builder.emplace_back(i);
      Solution::combine_helper(i + 1, n, k, builder, result);
      builder.pop_back();
    }
  }

public:
  std::vector<std::vector<int>> combine(int n, int k)
  {
    std::vector<int> builder;
    std::vector<std::vector<int>> result;
    Solution::combine_helper(1, n, k, builder, result);
    return result;
  }
};